storage/range.go

```diff
@@ -430,32 +504,41 @@ func (r *Range) addReadOnlyCmd(method string, args proto.Request, reply proto.Re
 
 	// If read-consistency is set to INCONSISTENT, run directly.
 	if header.ReadConsistency == proto.INCONSISTENT {
+		// But disallow any inconsistent reads within txns.
+		if args.Header().ReadConsistency == proto.INCONSISTENT && args.Header().Txn != nil {
+			reply.Header().SetGoError(util.Errorf("cannot allow inconsistent reads within a transaction"))
+			return reply.Header().GoError()
+		}
 		return r.executeCmd(0, method, args, reply)
+	} else if args.Header().ReadConsistency == proto.CONSENSUS {
+		reply.Header().SetGoError(util.Errorf("consensus reads not implemented"))
+		return reply.Header().GoError()
+	}
+
+	held, prevTerm := r.HasLeaderLease()
+	if !held {
+		// If we don't hold the lease, but nobody else does either, submit
+		// this read-only command as read-write to trigger an election.
+		if !r.AnotherHasLeaderLease() {
+			return r.addReadWriteCmd(method, args, reply, true)
+		}
+		reply.Header().SetGoError(r.newNotLeaderError())
```

Comment:
Is it a concern that this is not atomic? The leader lease could expire between the above AnotherHasLeaderLease and newNotLeaderError. It shouldn't really be a problem but it just looks a little fishy to examine the lease several times non-atomically.